dist/tools/firmware_metadata/generate-metadata.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2016, Mark Solters <msolters@gmail.com>.
+ *               2016, Inria
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+
+/**
+ * @ingroup     FW
+ * @file
+ * @brief       Meta-data generation for FW images
+ *
+ * @author      Mark Solters <msolters@gmail.com>
+ * @author      Francisco Acosta <francisco.acosta@inria.fr>
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/resource.h>
+
+#include "hashes/sha256.h"
+
+#define FW_METADATA_BIN "firmware-metadata.bin"
+
+typedef struct firmware_metadata {
+    uint8_t hash[SHA256_DIGEST_LENGTH];  /**< SHA256 Hash of firmware image */
+    uint8_t shash[SHA256_DIGEST_LENGTH]; /**< Signed SHA256 */
+    uint16_t version;                    /**< Integer representing firmware version */
+    uint32_t size;                       /**< Size of firmware image */
+    uint32_t appid;                      /**< Integer representing the application ID */
+} firmware_metadata_t;
+
+/* Input firmware .bin file */
+FILE *firmware_bin;
+
+/* Output metadata .bin file */
+FILE *metadata_bin;
+
+uint32_t firmware_size = 0;
```

Comment:
Does it matter for a native C program? In that case all the variables should be declared static in this program.